Level 3

pad1a opened the pull request:
Реализован функционал проекта согласно ТЗ

pad1a committed:
Написан класс карточки новости
Написан касс заполнения карточками блока с результатами поиска
Добавлен конфиг файл
Реализован поиск по запросу в форме поиска, при отсутствии результатов показывается блок "нет результатов"

pad1a committed:
Исправлена валидация пароля
Создан класс для работы с API Back

pad1a committed:
Передаем попапы в классы

pad1a committed:
Правки класса mainAPI

pad1a committed:
Обновление mainAPI

pad1a committed:
Устанавливаем параметр авториированного пользователя

pad1a committed:
Тесты с авторизацией к API

pad1a committed:
Работа с карточками новостей (добавление, удаление, отрисовка)

pad1a committed:
JS final v1

pad1a committed:
readme update

pad1a committed:
fix logout

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--8.0--> Настроена работа API для поиска новостей (https://newsapi.org/) и работы с сервером. Любые другие внешние API запрещены.
- Асинхронные запросы к API:
  - [x] <!--6.0--> запросы можно осуществлять через Fetch API или XMLHttpRequest; сторонние библиотеки запрещены — такие как axios или jQuery;
  - [x] <!--5.5--> для работы с API создан отдельный класс;
  - [ ] <!--5.5--> цепочка обработки промисов завершается блоком catch;
  **Комментарий: блоки catch расположены в методах класса Api, методы возвращают промис и выполнение цепочки then продолжается, получается catch в методах Api расположены в середине цепочки, они обрабатывают ошибку и выполнение продолжается как будто ошибки не было. Например при ошибке запроса данных в консоли ошибка перебора массива TypeError: Cannot read property 'length' of undefined. Т.е. несмотря на то что запрос выполнился неудачно программа попыталась отрисовать данные и упала уже на этом, такое поведение недопустимо. Поэтому необходимо убрать catch из методов Api и расположить их только в самом конце цепочки обработки промиса там где методы Api вызываются или если блок cacth находится в середине снова выбрасывать из него исключение или возвращать отлоненный промис**
  - [ ] <!--6.0--> первый обработчик then возрващает res.json;
  **Комментарий: если при запросе произошла ошибка должен возвращаться отклоненный промис, а не объект, см. пример в дп. клмментариях**
  - [ ] <!--6.0--> манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
  **Комментарий: карточки удаляются со страницы сохраненных до ответа сервера и даже если запрос выполнился успешно**
- Именование:
  - [ ] <!--2.5--> camelCase для параметров, методов и переменных;
  **Комментарий: каждое новое слово при camelCase должно быть с заглавной буквы, например mainapi newssum newstags newslist должно быть mainApi newsSum newsList. Проверить похожие места во всем проекте**
  - [x] <!--1.5--> существительные для имён переменных;
  - [ ] <!--1.5--> существительные во множественном числе для NodeList;
  **Комментарий:  if (buttonSave.length > 1) { if (buttonOutimage.length > 1) { if (nameEl.length > 1) { и т.д. , проверить во всем проекте - массивы нужно назвать переменные так, чтобы из их названия было понятно что это массив и какие элементы в нем хранятся (существительные во множественном числе)**  
  - [x] <!--2.5--> существительные с заглавной буквы для имён классов;
  - [x] <!--2.5--> const для неизменяемыx напрямую переменных;
  - [x] <!--1.5--> соответствующее содержимому имя класса;
  - [x] <!--1.5--> отсутсвует транслит и неуместные сокращения.
- [x] <!--6.0--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--3.5--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»
- [x] <!--8.0--> JWT сохраняется в браузере.
- [x] <!--6.0--> Код JS-модулей инкапсулирован.
- [x] <!--6.0--> Поля форм входа и регистрации валидируются на клиенте.

### Хорошие практики
- Код объектно-ориентирован:
  - [x] <!--0.7--> использованы ES6-классы;
  - [x] <!--1.1--> внутри классов нет создания экземпляров других классов;
  - [x] <!--0.7--> соблюдён принцип DRY (Don't Repeat Yourself);
  - [ ] <!--1.2--> соблюдён принцип единственной ответственности, то есть каждый класс полностью отвечает за свою задачу, и никакой другой класс к решению этой задачи не относится;
  **Комментарий: класс MainApi не должен участовать в изменении страницы, в нем должны описываться только запросы к серверу и возвращаться из методов промисы с данными**
  - [ ] <!--1.2--> соблюдён принцип открытости-закрытости: для добавления функциональности классы можно расширять, не модифицируя их;
  **Комментарий: в классе Popup захардкожены селекторы main_menu_button__auth popup__form_auth-link popup__form_newuser-link. Класс Popup должен быть универсальным для любого попапа. Сйчас же при добавлении попапа в проект придется менять код класса**
  - [x] <!--0.7--> у классов есть приватные поля, то есть свойства и методы, которые начинаются с нижнего подчёркивания и не используются нигде вне класса;
- Работа с DOM-деревом для вывода результатов:
  - [x] <!--1.2--> cайт защищён от XSS: нельзя выполнять скрипты через пользовательский ввод;
  - [x] <!--0.7--> нет потенциальных утечек памяти: обработчики созданы и удалены в нужных местах;
  - [x] <!--0.7--> NodeList перебирается с использованием методов для работы с массивами и псевдомассивами;
  - [ ] <!--0.7--> множественные операции c DOM-элементами производятся прежде, чем добавлять блок в разметку;
  **Комментарий: for переписать с использованием forEach или for of где это возможно**
  - [x] <!--0.7--> DOM-элементы не прерисовываются без необходимости;
  - [x] <!--1.2--> innerHTML запрещён — вместо него применяются более мягкие методы;
  - [x] <!--0.7--> поиск по DOM-дереву выполняется минимально возможное количество раз, то есть поиск одного элемента не выполняется дважды;
- Обработка ошибок API:
  - [ ] <!--1.2--> пользователь получает сообщение в случае ошибки;
  **Комментарий: при ошибке запроса новостей, регистрации или авторизации (например если нет интрента) должно быть собщение об ошибке**
  - [ ] <!--0.7--> поля формы заблокированы во время отправки запросов;
  **Комментарий: на время выполнения запроса нужно блокировать поля ввода и кнопку отправки формы**
- [x] <!--1.2--> Разные логические части приложения описаны в разных файлах и разнесены в разные папки. Интерфейсные классы, классы API, вспомогательные функции и константы отделены друг от друга.
- [x] <!--0.7--> При клике на «Выйти» страница автоматически меняет состояние и выглядит так, как у пользователя, не выполнившего вход в систему.

### Рекомендации
- [x] <!--0.6--> прелоудер крутится при ожидании ответа от сервера
- [x] <!--0.3--> `event.target` и `this` применяются правильно;
- Обработка и валидация форм:
  - [ ] <!--0.3--> текст ошибок кастомный, а не системный;
  **Комментарий: для поля ввода запроса новости также требуется кастомный текст ошибки если поле ввода пустое**
  - [x] <!--0.3--> валидация реализована и представляет собой отдельный метод для обработки форм
- Отсутствует «мусор» в коде:
  - [x] <!--0.3--> по имени колбэка слушателя понятно его назначение;
  - [ ] <!--0.3--> нет «магических значений», все значения записаны в переменные;
  **Комментарий: new Date(totimestamp - 7 * 24 * 3600 * 1000);**
  - [x] <!--0.3--> константы описаны в той области видимости, в которой объявлена функция, и внутри неё значения констант берутся из замыкания;
  - [ ] <!--0.3--> если с API можно работать несколькими способами, то выбран один способ, который применяется везде в проекте. Например, поиск отдельных элементов осуществляется только методом `querySelector`
  **Комментарий: вместо document.forms. и getElementById/ByClass нужно использовать querySelector**
- [ ] <!--0.3--> Любой попап закрывается не только при клике на иконку крестика, но и при клике в любое место вне попапа и при нажатии клавиши Esc.
**Комментарий: не закрывается по ESC**
- [x] <!--0.4--> После успешной регистрации пользователю показывается попап с сообщением: «Пользователь успешно зарегистрирован!» и предложением войти в систему
- [x] <!--0.4--> Если пользователь не вошёл в систему, то переход на страницу с сохранёнными новостями по прямой ссылке осуществляет редирект на главную.
- [x] <!--0.4--> Поля форм входа и регистрации валидируются на клиенте. Сообщения об ошибках кастомные, а не стандартные. Кастомные тексты ошибок есть в макете.
- [ ] <!--0.3--> Создан базовый компонент BaseComponent. Другие компоненты интерфейса наследуют у него.

## Количество баллов: 72

## Так же есть замечания по выполнению задания, без исправления которых, работа не будет принята, даже если наберет достаточное количество баллов
- при первом отображении карточек, должно отобразиться 3 карточки
- не работает кнопка "Показать ещё", при нажатии на неё должно отображаться дополнительно 3 новости, когда все новости отображены, кнопка "Показать ещё" должна скрываться
- при добавлении новости в сохраненные на её кнопке не отображается маркер сохраненного состояния


## Доп. Комментарии:
* при выполнении запроса на сервер нужна также проверка response.ok о том что запрос выполнился успешно
```
.then((res) => {    
	if (!res.ok) {     //если запрос выполнился неудачно возвращаем отклоненный промис
		return Promise.reject(`Ошибка: ${res.status}`); 
	}
	return res.json(); //если res.ok===true возвращаем результат запроса
})
```
- поле ввода пароля должно быть типа password
- проблемы с версткой 
![Screenshot_2](https://user-images.githubusercontent.com/9317133/88626425-45621280-d0b3-11ea-8a8b-866567fd7f3b.png)

pad1a committed:
Исправлено:
- camelCase для параметров, методов и переменных;
- существительные во множественном числе для NodeList;
- нет «магических значений», все значения записаны в переменные;

pad1a committed:
Исправлено (добавлен функционал):
- при первом отображении карточек, должно отобразиться 3 карточки;
- не работает кнопка "Показать ещё", при нажатии на неё должно отображаться дополнительно 3 новости, когда все новости отображены, кнопка "Показать ещё" должна скрываться;

pad1a committed:
Исправлено:
- при добавлении новости в сохраненные на её кнопке не отображается маркер сохраненного состояния.

Добавлена проверка на присутствие новости в сохраненных при рендеринге новостей на поиске, если уже сохранена то маркер подкрашивается.

pad1a committed:
Изменен адрес сервера API

pad1a committed:
Исправлено: манипуляции с DOM-элементами производятся после получения ответа сервера;
Карточки удаляются из сохраненных теперь после ответа сервера.

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--8.0--> Настроена работа API для поиска новостей (https://newsapi.org/) и работы с сервером. Любые другие внешние API запрещены.
- Асинхронные запросы к API:
  - [x] <!--6.0--> запросы можно осуществлять через Fetch API или XMLHttpRequest; сторонние библиотеки запрещены — такие как axios или jQuery;
  - [x] <!--5.5--> для работы с API создан отдельный класс;
  - [ ] <!--5.5--> цепочка обработки промисов завершается блоком catch;
  **Комментарий: блоки catch расположены в методах класса Api, методы возвращают промис и выполнение цепочки then продолжается, получается catch в методах Api расположены в середине цепочки, они обрабатывают ошибку и выполнение продолжается как будто ошибки не было. Например при ошибке запроса данных в консоли ошибка перебора массива TypeError: Cannot read property 'length' of undefined. Т.е. несмотря на то что запрос выполнился неудачно программа попыталась отрисовать данные и упала уже на этом, такое поведение недопустимо. Поэтому необходимо убрать catch из методов Api и расположить их только в самом конце цепочки обработки промиса там где методы Api вызываются или если блок cacth находится в середине снова выбрасывать из него исключение или возвращать отлоненный промис. При использовании async/await (let res = await promise;) должна быть обработка ошибок try/catch**
  - [ ] <!--6.0--> первый обработчик then возрващает res.json;
  **Комментарий: если при запросе произошла ошибка должен возвращаться отклоненный промис, а не объект, см. пример в дп. комментариях**
  - [ ] <!--6.0--> манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
  **Комментарий: при добавлении карточки в сохраненные, маркер устанавливается до ответа сервера. Карточки удаляются из сохраненных даже если запрос выполнился с ошибкой**
- Именование:
  - [x] <!--2.5--> camelCase для параметров, методов и переменных;
  - [x] <!--1.5--> существительные для имён переменных;
  - [x] <!--1.5--> существительные во множественном числе для NodeList;
 
  - [x] <!--2.5--> существительные с заглавной буквы для имён классов;
  - [x] <!--2.5--> const для неизменяемыx напрямую переменных;
  - [x] <!--1.5--> соответствующее содержимому имя класса;
  - [x] <!--1.5--> отсутсвует транслит и неуместные сокращения.
- [x] <!--6.0--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--3.5--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»
- [x] <!--8.0--> JWT сохраняется в браузере.
- [x] <!--6.0--> Код JS-модулей инкапсулирован.
- [x] <!--6.0--> Поля форм входа и регистрации валидируются на клиенте.

### Хорошие практики
- Код объектно-ориентирован:
  - [x] <!--0.7--> использованы ES6-классы;
  - [x] <!--1.1--> внутри классов нет создания экземпляров других классов;
  - [x] <!--0.7--> соблюдён принцип DRY (Don't Repeat Yourself);
  - [ ] <!--1.2--> соблюдён принцип единственной ответственности, то есть каждый класс полностью отвечает за свою задачу, и никакой другой класс к решению этой задачи не относится;
  **Комментарий: класс MainApi не должен участовать в изменении страницы, в нем должны описываться только запросы к серверу и возвращаться из методов промисы с данными**
  - [ ] <!--1.2--> соблюдён принцип открытости-закрытости: для добавления функциональности классы можно расширять, не модифицируя их;
  **Комментарий: в классе Popup захардкожены селекторы main_menu_button__auth popup__form_auth-link popup__form_newuser-link. Класс Popup должен быть универсальным для любого попапа. Сйчас же при добавлении попапа в проект придется менять код класса**
  - [x] <!--0.7--> у классов есть приватные поля, то есть свойства и методы, которые начинаются с нижнего подчёркивания и не используются нигде вне класса;
- Работа с DOM-деревом для вывода результатов:
  - [x] <!--1.2--> cайт защищён от XSS: нельзя выполнять скрипты через пользовательский ввод;
  - [x] <!--0.7--> нет потенциальных утечек памяти: обработчики созданы и удалены в нужных местах;
  - [x] <!--0.7--> NodeList перебирается с использованием методов для работы с массивами и псевдомассивами;
  - [ ] <!--0.7--> множественные операции c DOM-элементами производятся прежде, чем добавлять блок в разметку;
  **Комментарий: for переписать с использованием forEach или for of где это возможно**
  - [x] <!--0.7--> DOM-элементы не прерисовываются без необходимости;
  - [x] <!--1.2--> innerHTML запрещён — вместо него применяются более мягкие методы;
  - [x] <!--0.7--> поиск по DOM-дереву выполняется минимально возможное количество раз, то есть поиск одного элемента не выполняется дважды;
- Обработка ошибок API:
  - [ ] <!--1.2--> пользователь получает сообщение в случае ошибки;
  **Комментарий: при ошибке запроса новостей, регистрации или авторизации (например если нет интрента) должно быть собщение об ошибке**
  - [ ] <!--0.7--> поля формы заблокированы во время отправки запросов;
  **Комментарий: на время выполнения запроса нужно блокировать поля ввода и кнопку отправки формы**
- [x] <!--1.2--> Разные логические части приложения описаны в разных файлах и разнесены в разные папки. Интерфейсные классы, классы API, вспомогательные функции и константы отделены друг от друга.
- [x] <!--0.7--> При клике на «Выйти» страница автоматически меняет состояние и выглядит так, как у пользователя, не выполнившего вход в систему.

### Рекомендации
- [x] <!--0.6--> прелоудер крутится при ожидании ответа от сервера
- [x] <!--0.3--> `event.target` и `this` применяются правильно;
- Обработка и валидация форм:
  - [ ] <!--0.3--> текст ошибок кастомный, а не системный;
  **Комментарий: для поля ввода запроса новости также требуется кастомный текст ошибки если поле ввода пустое**
  - [x] <!--0.3--> валидация реализована и представляет собой отдельный метод для обработки форм
- Отсутствует «мусор» в коде:
  - [x] <!--0.3--> по имени колбэка слушателя понятно его назначение;
  - [x] <!--0.3--> нет «магических значений», все значения записаны в переменные;
  - [x] <!--0.3--> константы описаны в той области видимости, в которой объявлена функция, и внутри неё значения констант берутся из замыкания;
  - [ ] <!--0.3--> если с API можно работать несколькими способами, то выбран один способ, который применяется везде в проекте. Например, поиск отдельных элементов осуществляется только методом `querySelector`
  **Комментарий: вместо document.forms. и getElementById/ByClass нужно использовать querySelector**
- [ ] <!--0.3--> Любой попап закрывается не только при клике на иконку крестика, но и при клике в любое место вне попапа и при нажатии клавиши Esc.
**Комментарий: не закрывается по ESC**
- [x] <!--0.4--> После успешной регистрации пользователю показывается попап с сообщением: «Пользователь успешно зарегистрирован!» и предложением войти в систему
- [x] <!--0.4--> Если пользователь не вошёл в систему, то переход на страницу с сохранёнными новостями по прямой ссылке осуществляет редирект на главную.
- [x] <!--0.4--> Поля форм входа и регистрации валидируются на клиенте. Сообщения об ошибках кастомные, а не стандартные. Кастомные тексты ошибок есть в макете.
- [ ] <!--0.3--> Создан базовый компонент BaseComponent. Другие компоненты интерфейса наследуют у него.

## Количество баллов: 76

## Так же есть замечания по выполнению задания, без исправления которых, работа не будет принята, даже если наберет достаточное количество баллов
- при отображении всех новостей, кнопка "Показать ещё" не скрывается, а при нажатии начинаю падать ошибки "Uncaught TypeError: Cannot read property 'classList' of undefined"

pad1a committed:
Исправлено:
при отображении всех новостей, кнопка "Показать ещё" не скрывается, а при нажатии начинаю падать ошибки "Uncaught TypeError: Cannot read property 'classList' of undefined"

pad1a committed:
Исправлено:
манипуляции с DOM-элементами
при добавлении карточки в сохраненные, маркер устанавливается до ответа сервера. Карточки удаляются из сохраненных даже если запрос выполнился с ошибкой

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--8.0--> Настроена работа API для поиска новостей (https://newsapi.org/) и работы с сервером. Любые другие внешние API запрещены.
- Асинхронные запросы к API:
  - [x] <!--6.0--> запросы можно осуществлять через Fetch API или XMLHttpRequest; сторонние библиотеки запрещены — такие как axios или jQuery;
  - [x] <!--5.5--> для работы с API создан отдельный класс;
  - [ ] <!--5.5--> цепочка обработки промисов завершается блоком catch;
  **Комментарий: блоки catch расположены в методах класса Api, методы возвращают промис и выполнение цепочки then продолжается, получается catch в методах Api расположены в середине цепочки, они обрабатывают ошибку и выполнение продолжается как будто ошибки не было. Например при ошибке запроса данных в консоли ошибка перебора массива TypeError: Cannot read property 'length' of undefined. Т.е. несмотря на то что запрос выполнился неудачно программа попыталась отрисовать данные и упала уже на этом, такое поведение недопустимо. Поэтому необходимо убрать catch из методов Api и расположить их только в самом конце цепочки обработки промиса там где методы Api вызываются или если блок cacth находится в середине снова выбрасывать из него исключение или возвращать отлоненный промис. При использовании async/await (let res = await promise;) должна быть обработка ошибок try/catch**
  - [ ] <!--6.0--> первый обработчик then возрващает res.json;
  **Комментарий: если при запросе произошла ошибка должен возвращаться отклоненный промис, а не объект, см. пример в дп. комментариях**
  - [x] <!--6.0--> манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
- Именование:
  - [x] <!--2.5--> camelCase для параметров, методов и переменных;
  - [x] <!--1.5--> существительные для имён переменных;
  - [x] <!--1.5--> существительные во множественном числе для NodeList;
  - [x] <!--2.5--> существительные с заглавной буквы для имён классов;
  - [x] <!--2.5--> const для неизменяемыx напрямую переменных;
  - [x] <!--1.5--> соответствующее содержимому имя класса;
  - [x] <!--1.5--> отсутсвует транслит и неуместные сокращения.
- [x] <!--6.0--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--3.5--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»
- [x] <!--8.0--> JWT сохраняется в браузере.
- [x] <!--6.0--> Код JS-модулей инкапсулирован.
- [x] <!--6.0--> Поля форм входа и регистрации валидируются на клиенте.

### Хорошие практики
- Код объектно-ориентирован:
  - [x] <!--0.7--> использованы ES6-классы;
  - [x] <!--1.1--> внутри классов нет создания экземпляров других классов;
  - [x] <!--0.7--> соблюдён принцип DRY (Don't Repeat Yourself);
  - [ ] <!--1.2--> соблюдён принцип единственной ответственности, то есть каждый класс полностью отвечает за свою задачу, и никакой другой класс к решению этой задачи не относится;
  **Комментарий: класс MainApi не должен участовать в изменении страницы, в нем должны описываться только запросы к серверу и возвращаться из методов промисы с данными**
  - [ ] <!--1.2--> соблюдён принцип открытости-закрытости: для добавления функциональности классы можно расширять, не модифицируя их;
  **Комментарий: в классе Popup захардкожены селекторы main_menu_button__auth popup__form_auth-link popup__form_newuser-link. Класс Popup должен быть универсальным для любого попапа. Сйчас же при добавлении попапа в проект придется менять код класса**
  - [x] <!--0.7--> у классов есть приватные поля, то есть свойства и методы, которые начинаются с нижнего подчёркивания и не используются нигде вне класса;
- Работа с DOM-деревом для вывода результатов:
  - [x] <!--1.2--> cайт защищён от XSS: нельзя выполнять скрипты через пользовательский ввод;
  - [x] <!--0.7--> нет потенциальных утечек памяти: обработчики созданы и удалены в нужных местах;
  - [x] <!--0.7--> NodeList перебирается с использованием методов для работы с массивами и псевдомассивами;
  - [ ] <!--0.7--> множественные операции c DOM-элементами производятся прежде, чем добавлять блок в разметку;
  **Комментарий: for переписать с использованием forEach или for of где это возможно**
  - [x] <!--0.7--> DOM-элементы не прерисовываются без необходимости;
  - [x] <!--1.2--> innerHTML запрещён — вместо него применяются более мягкие методы;
  - [x] <!--0.7--> поиск по DOM-дереву выполняется минимально возможное количество раз, то есть поиск одного элемента не выполняется дважды;
- Обработка ошибок API:
  - [ ] <!--1.2--> пользователь получает сообщение в случае ошибки;
  **Комментарий: при ошибке запроса новостей, регистрации или авторизации (например если нет интрента) должно быть собщение об ошибке**
  - [ ] <!--0.7--> поля формы заблокированы во время отправки запросов;
  **Комментарий: на время выполнения запроса нужно блокировать поля ввода и кнопку отправки формы**
- [x] <!--1.2--> Разные логические части приложения описаны в разных файлах и разнесены в разные папки. Интерфейсные классы, классы API, вспомогательные функции и константы отделены друг от друга.
- [x] <!--0.7--> При клике на «Выйти» страница автоматически меняет состояние и выглядит так, как у пользователя, не выполнившего вход в систему.

### Рекомендации
- [x] <!--0.6--> прелоудер крутится при ожидании ответа от сервера
- [x] <!--0.3--> `event.target` и `this` применяются правильно;
- Обработка и валидация форм:
  - [ ] <!--0.3--> текст ошибок кастомный, а не системный;
  **Комментарий: для поля ввода запроса новости также требуется кастомный текст ошибки если поле ввода пустое**
  - [x] <!--0.3--> валидация реализована и представляет собой отдельный метод для обработки форм
- Отсутствует «мусор» в коде:
  - [x] <!--0.3--> по имени колбэка слушателя понятно его назначение;
  - [x] <!--0.3--> нет «магических значений», все значения записаны в переменные;
  - [x] <!--0.3--> константы описаны в той области видимости, в которой объявлена функция, и внутри неё значения констант берутся из замыкания;
  - [ ] <!--0.3--> если с API можно работать несколькими способами, то выбран один способ, который применяется везде в проекте. Например, поиск отдельных элементов осуществляется только методом `querySelector`
  **Комментарий: вместо document.forms. и getElementById/ByClass нужно использовать querySelector**
- [ ] <!--0.3--> Любой попап закрывается не только при клике на иконку крестика, но и при клике в любое место вне попапа и при нажатии клавиши Esc.
**Комментарий: не закрывается по ESC**
- [x] <!--0.4--> После успешной регистрации пользователю показывается попап с сообщением: «Пользователь успешно зарегистрирован!» и предложением войти в систему
- [x] <!--0.4--> Если пользователь не вошёл в систему, то переход на страницу с сохранёнными новостями по прямой ссылке осуществляет редирект на главную.
- [x] <!--0.4--> Поля форм входа и регистрации валидируются на клиенте. Сообщения об ошибках кастомные, а не стандартные. Кастомные тексты ошибок есть в макете.
- [ ] <!--0.3--> Создан базовый компонент BaseComponent. Другие компоненты интерфейса наследуют у него.

## Количество баллов: 82

## Поздравляю с завершением дипломного проекта!

### Советую все же не забрасывать разработку, чтобы обучение не прошло зря т.к. полученные знания быстро забываются. Несколько направлений для дальнейшего развития:
- свертайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- изучить фреймворк - React или Vue.js, хотя бы на базовом уровне
- ходить на собеседования, опыт прохождения собеседований не менее важен знаний теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте habr и слушаете подкаксты:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw
